services/rollout-service/pkg/argo/argo.go

```diff
@@ -0,0 +1,341 @@
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+MIT License for more details.
+
+You should have received a copy of the MIT License
+along with kuberpult. If not, see <https://directory.fsf.org/wiki/License:Expat>.
+
+Copyright 2023 freiheit.com
+*/
+package argo
+
+import (
+	"context"
+	"fmt"
+	"github.com/argoproj/argo-cd/v2/pkg/apiclient/application"
+	"github.com/argoproj/argo-cd/v2/pkg/apis/application/v1alpha1"
+	"github.com/freiheit-com/kuberpult/pkg/api"
+	"github.com/freiheit-com/kuberpult/pkg/logger"
+	"github.com/freiheit-com/kuberpult/pkg/ptr"
+	"github.com/freiheit-com/kuberpult/pkg/setup"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"path/filepath"
+	"slices"
+)
+
+// this is a simpler version of ApplicationServiceClient from the application package
+type SimplifiedApplicationServiceClient interface {
+	Watch(ctx context.Context, qry *application.ApplicationQuery, opts ...grpc.CallOption) (application.ApplicationService_WatchClient, error)
+}
+
+type ArgoAppProcessor struct {
+	trigger               chan *api.GetOverviewResponse
+	lastOverview          *api.GetOverviewResponse
+	argoApps              chan *v1alpha1.ApplicationWatchEvent
+	ApplicationClient     application.ApplicationServiceClient
+	ManageArgoAppsEnabled bool
+	ManageArgoAppsFilter  []string
+}
+
+func New(appClient application.ApplicationServiceClient, manageArgoApplicationEnabled bool, manageArgoApplicationFilter []string) ArgoAppProcessor {
+	return ArgoAppProcessor{
+		ApplicationClient:     appClient,
+		ManageArgoAppsEnabled: manageArgoApplicationEnabled,
+		ManageArgoAppsFilter:  manageArgoApplicationFilter,
+		trigger:               make(chan *api.GetOverviewResponse),
+		argoApps:              make(chan *v1alpha1.ApplicationWatchEvent),
+	}
+}
+
+type Key struct {
+	AppName     string
+	EnvName     string
+	Application *api.Environment_Application
+	Environment *api.Environment
+}
+
+func (a *ArgoAppProcessor) Push(ctx context.Context, last *api.GetOverviewResponse) {
+	l := logger.FromContext(ctx).With(zap.String("argo-pushing", "ready"))
+	a.lastOverview = last
+	select {
+	case a.trigger <- a.lastOverview:
+		l.Info("argocd.pushed")
+	default:
+	}
+}
+
+func (a *ArgoAppProcessor) Consume(ctx context.Context) error {
+	l := logger.FromContext(ctx).With(zap.String("event-consuming", "ready"))
+
+	for {
+		select {
+		case overview := <-a.trigger:
+			for _, envGroup := range overview.EnvironmentGroups {
+				for _, env := range envGroup.Environments {
+					appList, err := a.ApplicationClient.List(ctx, &application.ApplicationQuery{
+						Name: ptr.FromString(fmt.Sprintf("%s-", env)),
+					})
+
+					if err != nil {
+						fmt.Errorf("listing applications: %w", err)
+					}
+					err = a.DeleteArgoApps(ctx, appList, env.Applications)
```

Comment:
It's a bit counter-intuitive to call "delete" when "list" didn't work.
How about this:
```suggestion
						fmt.Errorf("listing applications: %w", err)
						continue
					}
					err = a.DeleteArgoApps(ctx, appList, env.Applications)
```